perf/pkeyread.c

```diff
@@ -0,0 +1,381 @@
+    BIO *pem;
+    int i;
+    size_t len;
+    OSSL_TIME start, end;
+
+    if (sample_id >= SAMPLE_ALL) {
+        fprintf(stderr, "%s no sample key set for test\n", __func__);
+        err = 1;
+        return;
+    }
+
+    keydata = sample_keys[sample_id][FORMAT_PEM];
+    keydata_sz = sample_key_sizes[sample_id][FORMAT_PEM];
+    pem = BIO_new_mem_buf(keydata, keydata_sz);
+
+    if (pem == NULL) {
+        fprintf(stderr, "%s Cannot create mem BIO [%s PEM]\n",
+            __func__, sample_names[sample_id]);
+        err = 1;
+        return;
+    }
+
+    start = ossl_time_now();
+
+    /*
+     * Technically this includes the EVP_PKEY_free() in the timing - but I
+     * think we can live with that
+     */
+    for (i = 0; i < NUM_CALLS_PER_TEST / threadcount; i++) {
+        key = PEM_read_bio_PrivateKey(pem, NULL, NULL, NULL);
+        if (key == NULL) {
+            fprintf(stderr, "Failed to create key: %d [%s PEM]\n", i,
+                sample_names[sample_id]);
+            err = 1;
+            BIO_free(pem);
+            return;
+        }
+        EVP_PKEY_free(key);
+        BIO_reset(pem);
+    }
+
+    end = ossl_time_now();
+    times[num] = ossl_time_subtract(end, start);
+
+    BIO_free(pem);
+}
+
+void do_derread(size_t num)
+{
+    const char *keydata;
+    size_t keydata_sz;
+    const unsigned char *pk_buf = NULL;
+    int pk_buf_len;
+    BIO *der;
+    PKCS8_PRIV_KEY_INFO *pkey_info8 = NULL;
+    EVP_PKEY *pkey = NULL;
+    X509_PUBKEY *key = NULL;
+    int i;
+    OSSL_TIME start, end;
+
+    if (sample_id >= SAMPLE_ALL) {
+        fprintf(stderr, "%s no sample key set for test\n", __func__);
+        err = 1;
+        return;
+    }
+
+    keydata = sample_keys[sample_id][FORMAT_DER];
+    keydata_sz = sample_key_sizes[sample_id][FORMAT_DER];
+    der = BIO_new_mem_buf(keydata, keydata_sz);
+    if (der == NULL) {
+        fprintf(stderr, "%s Cannot create mem BIO [%s DER]\n",
+            __func__, sample_names[sample_id]);
+        err = 1;
+        return;
+    }
+
+    start = ossl_time_now();
+
+    for (i = 0; i < NUM_CALLS_PER_TEST / threadcount && err == 0; i++) {
+        pkey_info8 = d2i_PKCS8_PRIV_KEY_INFO_bio(der, &pkey_info8);
+        if (pkey_info8 == NULL) {
+            fprintf(stderr, "%s pkey_info8 is NULL [%s BER]\n",
+                __func__, sample_names[sample_id]);
+            err = 1;
+            goto error;
+        }
+
+        pkey = EVP_PKCS82PKEY(pkey_info8);
+        if (pkey == NULL) {
+            fprintf(stderr, "%s pkey is NULL [%s BER]\n",
+                __func__, sample_names[sample_id]);
+            err = 1;
+            goto error;
+        }
+
+        if (X509_PUBKEY_set(&key, pkey) == 0) {
+            fprintf(stderr, "%s X509_PUBKEY_set() failed [%s BER]\n",
+                __func__, sample_names[sample_id]);
+            err = 1;
+            goto error;
+        }
+
+        if (X509_PUBKEY_get0_param(NULL, &pk_buf, &pk_buf_len, NULL, key) == 0) {
```

Comment:
Not really sure why we do the X509_PUBKEY_set and X509_PUBKEY_get0_param calls here. This is primarily intended as a decoder performance test. But these calls don't have much to do with decoding.

Reply:
I took them from code sample which is referenced here:
https://gist.github.com/richsalz/31a98a3095fa36ab6a66082a0c028a6f

do you want me to remove calls to `X509_PUBKEY_*()` functions?

Reply:
Yes, I would remove them. They do not seem relevant to the main point of the test.